### PR TITLE
Standard provider bash operator

### PR DESCRIPTION
**Pull request by gopidesupavan**
Moving bash operators/sensors to standard provider

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by gopidesupavan** [comment]
moving this bash operator to standard provider, has module path changes in systems test, integration tests.

failures in python client tests, i am thinking due to bashoperator path change, causing dag not loaded.
Checking remaining tests.

File: .pre-commit-config.yaml
Comment (jscheffl) [inline .pre-commit-config.yaml]: For this change - if we basically add all example DAGs into the exclude list - the check has almost no effect.

Can you re-work the check such that no provider except standard provider dependencies are accepted?

File: .pre-commit-config.yaml
Comment (jscheffl) [inline .pre-commit-config.yaml]: Why does this need an exclusion? This is not a system test. Does the check raise a false-positive and need a fix?

File: airflow/example_dags/example_bash_operator.py
Comment (jscheffl) [inline airflow/example_dags/example_bash_operator.py:1]: This all looks a bit strange now in my eyes. All example DAGs sitting in the core are requiring a provider package. Smells like a cyclomatic package dependency.

1) Should we move the example DAGs into the provider as well? (Then they are hard to find :-( )
2) Should we load  the core operators from the compatability alias? (Then we have bad examples)
3) Any other ideas?

File: .pre-commit-config.yaml
Comment (gopidesupavan) [reply .pre-commit-config.yaml]: Yes, agree. will check.

File: .pre-commit-config.yaml
Comment (gopidesupavan) [reply .pre-commit-config.yaml]: Good catch thank you, have removed that from exclusions.

File: airflow/example_dags/example_bash_operator.py
Comment (jscheffl) [reply airflow/example_dags/example_bash_operator.py:1]: Note: This was an open question to the community :-D

Comment (jedcunningham) [reply airflow/example_dags/example_bash_operator.py:1]: 1. Seems reasonable to me. Not that hard to find. But `load_examples` probably needs more work then. Probably not be worth it.
2. No.
3. Leave it? Airflow 3 will require the standard provider anyways right?

Side note: We don't need the try/except fallback imo.

File: airflow/providers/standard/provider.yaml
Comment (jedcunningham) [inline airflow/providers/standard/provider.yaml]: My 2c, I'm really not sure we need the time/core part of this. Seems better to just have

```
airflow.providers.standard.operators.weekday
airflow.providers.standard.operators.bash
```

Comment (gopidesupavan) [reply airflow/providers/standard/provider.yaml]: i am ok with suggestion. :)

File: airflow/example_dags/example_bash_operator.py
Comment (ashb) [reply airflow/example_dags/example_bash_operator.py:1]: What do we do for the other bundled providers - http, sqlite (plus two others - ftp and email)?

**Comment by gopidesupavan** [comment]
Could you please help on this, kubernetes executor tests are failing with dag not found `example_kubernetes_executor ` and `example_bash_operator`. one option i tried copying the standard package to k8s build container, but its still failing with same reason.

**Comment by gopidesupavan** [comment]
Additional prod image tests also failing, as it doesnt have the standard provider. inside latest prod image pull.
image

**Comment by gopidesupavan** [comment]
> Additional prod image tests also failing, as it doesnt have the standard provider. inside latest prod image pull. image

Any suggestions for these tests?

**Comment by gopidesupavan** [comment]
After reviewing the logs, it seems the reason the k8s tests are failing is due to the absence of the standard provider package in the image. This is because the standard provider not included in the prod_image_installed_providers.txt file.

Is it okay to add standard provider in to that file?

**Comment by jscheffl** [comment]
> After reviewing the logs, it seems the reason the k8s tests are failing is due to the absence of the standard provider package in the image. This is because i guess the standard provider not included in the prod_image_installed_providers.txt file ?
> 
> Is it okay to add standard provider in to that file?

Yes. At least from the community discussion we always expected that the new "standard" provider will be a standard install and dependency. So adding to prod image would be the natural consequence. Hope that fixes the problems you had (and sorry for responding late,,,)

**Comment by gopidesupavan** [comment]
> > After reviewing the logs, it seems the reason the k8s tests are failing is due to the absence of the standard provider package in the image. This is because i guess the standard provider not included in the prod_image_installed_providers.txt file ?
> > Is it okay to add standard provider in to that file?
> 
> Yes. At least from the community discussion we always expected that the new "standard" provider will be a standard install and dependency. So adding to prod image would be the natural consequence. Hope that fixes the problems you had (and sorry for responding late,,,)

Thank you :) , I am glad the debug helped me to understand how the ci workflow setups and working process.

File: .pre-commit-config.yaml
Comment (gopidesupavan) [reply .pre-commit-config.yaml]: updated to allow only imports from standard provider :)

File: airflow/providers/standard/provider.yaml
Comment (gopidesupavan) [reply airflow/providers/standard/provider.yaml]: removed core/time package folders from standard provider.

**Comment by gopidesupavan** [comment]
Added standard provider to prod_image_installed_providers, but the CI pipeline is ignoring, might be its not in ready state to consider for constraints while building image? any suggestions.

File: airflow/example_dags/example_bash_operator.py
Comment (gopidesupavan) [reply airflow/example_dags/example_bash_operator.py:1]: If this refactor is necessary now, I am happy to do part of this or new pr. Please kindly suggest?

Comment (jscheffl) [reply airflow/example_dags/example_bash_operator.py:1]: Okay, yes - I would agree that a refactor of the examples might be possible in a separate PR as well - would not change the scope or structure of the Standard provider.

**Comment by gopidesupavan** [comment]
The standard provider hasn't been released yet, which is why it's not being installed in the production build images. As a result, k8s and one Docker test are failing. As a temporary, I’ve updated these tests to be skipped until the standard provider is released and added to prod_image_installed_providers.txt. Could you please let me know if this approach is okay or if you have any other suggestions? :)

I also tried updating prod_image_installed_providers.txt to check if it would install, but it’s still not working.

File: .pre-commit-config.yaml
Comment (gopidesupavan) [reply .pre-commit-config.yaml]: Hope the updated check is fine?

File: airflow/providers/openlineage/provider.yaml
Comment (gopidesupavan) [inline airflow/providers/openlineage/provider.yaml:87]: I am not sure about this change impact, will it cause any issues? when this provider release in the next provider release cycle, without standard provider? could you please give your thoughts?

Comment (jscheffl) [reply airflow/providers/openlineage/provider.yaml:87]: @eladkal This in-deed is a chicken-and-egg problem we had last time with FAB as well. How about if we make a "small" PR such that the standard provoder is ready, release this once (manually) to have it as an empty or half-bake version such that we can un-block this loop?

Comment (gopidesupavan) [reply airflow/providers/openlineage/provider.yaml:87]: Thanks jens.

Yes small release will help us break the loop, so there's no need to skip the k8s and Docker tests that were recently marked as skipped in the latest commit.

Additionally, hooks related to FileHook and PackageIndexHook are being used in the provider manager, which could also cause test failures when moved to the standard provider.

Comment (gopidesupavan) [reply airflow/providers/openlineage/provider.yaml:87]: Hi @eladkal any thoughts on this please?

Comment (ashb) [reply airflow/providers/openlineage/provider.yaml:87]: Why do we need a release? We can add it to the standard list well before we do a release of providers.

Comment (ashb) [reply airflow/providers/openlineage/provider.yaml:87]: > Additionally, hooks related to FileHook and PackageIndexHook are being used in the provider manager, which could also cause test failures when moved to the standard provider

@jscheffl You added this -- looking at the code (but not the PR desc, forgive me if it's mentioned there) why does the provider manager need to import these two classes eagerly the way it does?

Comment (gopidesupavan) [reply airflow/providers/openlineage/provider.yaml:87]: sorry i could have referenced code links:

https://github.com/apache/airflow/blob/main/airflow/providers_manager.py#L39
https://github.com/apache/airflow/blob/main/airflow/providers_manager.py#L40

Comment (gopidesupavan) [reply airflow/providers/openlineage/provider.yaml:87]: > Why do we need a release? We can add it to the standard list well before we do a release of providers.

@ashb i have tried adding standard provider to prod_image_installed_providers.txt , but this package is not getting installed in container image, which cause the test failures in here `kubernetes_tests/test_kubernetes_executor.py` and `docker_tests/test_docker_compose_quick_start.py`. 

celery executor utils is using bash/python operator imports:

https://github.com/apache/airflow/pull/42252/files#diff-0e0d141920152b568dd569500b73ebaca58893fa2a98255ef5a565743f5e374aR119

As a work around i have added skip check, i feel this is not good idea to skip these tests , may be you can have more view on these , whether the skip correct way to solve these test failures. 

https://github.com/apache/airflow/pull/42252/commits/bd03162e007be7268b0ebb7b650887ea5f9141fc#diff-8866ef4555c16e5398e9e5f7194c3b3251c69e2adba9b57bc899d421305d0b17R31

https://github.com/apache/airflow/pull/42252/commits/b553358ead17b768c03ec1c3a895a6b4520a0aeb#diff-0f21d6ea329e8ba2276c0c2c945d7325ac64e6dae24f27a858c7cf0fab6760e5R70

Comment (jscheffl) [reply airflow/providers/openlineage/provider.yaml:87]: > > Additionally, hooks related to FileHook and PackageIndexHook are being used in the provider manager, which could also cause test failures when moved to the standard provider
> 
> @jscheffl You added this -- looking at the code (but not the PR desc, forgive me if it's mentioned there) why does the provider manager need to import these two classes eagerly the way it does?

@ashb ... don't know... did I add the imports? I don't know the reason or we never had the problem. If we make the import lazy and this would fix the problem I am okay. But also providers manager is loading a lot of stuff during start, am not sure whether this resolves or just makes the problems a step later. It is worth trying to lazy import. Might even improve startup time as side effect if we are lucky :-D

Comment (potiuk) [reply airflow/providers/openlineage/provider.yaml:87]: See the devlist message that I sent https://lists.apache.org/thread/1gshy5cjmp9wz5v8dozyh64jj3dyn5s4 - you cannot add standard provider to "prod_image_installed_providers.txt" because it's not released in PyPI (yet) - but we can release it's pre-release version manually.

Comment (potiuk) [reply airflow/providers/openlineage/provider.yaml:87]: FYI: We do not need to switch "not-read" temporarily for that - we can use direct breeze command to build the package with .dev0 version and upload it to PyPI. I think that's the easiest way.

This is as simple as:

```bash
breeze release-management prepare-provider-packages standard --version-suffix-for-pypi dev0
```

Followed by:

```bash
twine upload dist/apache_airflow_providers_standard-1.0.0.dev0-py3-none-any.whl
```

The only prerequisite is that person doing it has PyPI credentials configured.

This way we can upload it any time without waiting for regular release process - i.e. when we move some stuff to standard provider, we can upload it to PyPI from the PR branch, before we merge the PR.

Comment (gopidesupavan) [reply airflow/providers/openlineage/provider.yaml:87]: Thank you for this, I will remove the skipif checks added to k8s and docker tests on this pr.

Comment (potiuk) [reply airflow/providers/openlineage/provider.yaml:87]: > I will remove the skipif checks added to k8s and docker tests on this pr.

Sure, We will need to release the "dev0" version of providers first after we get to consensus that this is the right thing to do.

Comment (potiuk) [reply airflow/providers/openlineage/provider.yaml:87]: OK. I uploaded the standard package .dev0 to try it - https://pypi.org/project/apache-airflow-providers-standard/1.0.0.dev0/ -> you can try to remove the skip-if in your PR now @gopidesupavan and we will take it from there.

Comment (gopidesupavan) [reply airflow/providers/openlineage/provider.yaml:87]: Thank you :) updated now.

Comment (gopidesupavan) [reply airflow/providers/openlineage/provider.yaml:87]: @potiuk it looks like not adding to container image https://github.com/apache/airflow/actions/runs/11159243493/job/31018063584?pr=42252#step:12:606 , any constrain file changes required? 

failure here: https://github.com/apache/airflow/actions/runs/11159243533/job/31018299717?pr=42252#step:8:1284

Comment (jscheffl) [reply airflow/providers/openlineage/provider.yaml:87]: Might it be related to https://pip.pypa.io/en/stable/cli/pip_install/#pre-release-versions and we need to pin specifically to the dev version to have it installed?

Comment (gopidesupavan) [reply airflow/providers/openlineage/provider.yaml:87]: not sure where to update this? any help please.

Comment (jscheffl) [reply airflow/providers/openlineage/provider.yaml:87]: I am not 100% sure and we never had this use case before but I assume it must be changed in hatch_build.py:52 and there the currect dev version must be pinned. (Like the version of FAB also is in line 42)

Comment (gopidesupavan) [reply airflow/providers/openlineage/provider.yaml:87]: Thanks jens, updated let me check if this picks up or not.

Comment (gopidesupavan) [reply airflow/providers/openlineage/provider.yaml:87]: it didnt installed. trying another approach, gone through the prod-image-build file, there are steps installing the packages, added extra step to install standard provider, let me check if this installs package or not.

Comment (gopidesupavan) [reply airflow/providers/openlineage/provider.yaml:87]: No luck, even updated CHICKEN_EGG_PROVIDERS in global constants, build is not considering this.

Comment (jscheffl) [reply airflow/providers/openlineage/provider.yaml:87]: Mhm... no idea... @potiuk I assume we need your master expert expertise :-D

Comment (jscheffl) [reply airflow/providers/openlineage/provider.yaml:87]: Note that the failed backcompat in the last build is (hopefully) fixed in https://github.com/apache/airflow/pull/42750 .. the other might be an in-stability... maybe we need to re-base once the main is green as well.

Comment (gopidesupavan) [reply airflow/providers/openlineage/provider.yaml:87]: Woohooo finally all the tests went through green , except one but this is not related to this change. 😄 

Thanks all for the help....

**Review by jscheffl — APPROVED** [review]
Thanks for the endurance to make this finally green!

As this is a critical thing, before merge, can we have a second pair of eyes for review? Good from my side

**Review by romsharon98 — APPROVED** [review]

**Comment by jscheffl** [comment]
Oh, nooo, merge conflict - can you fix this? then LGTM!

**Comment by gopidesupavan** [comment]
> Oh, nooo, merge conflict - can you fix this? then LGTM!

yeah just resolved 3.8 removals 😄

**Comment by gopidesupavan** [comment]
> > Oh, nooo, merge conflict - can you fix this? then LGTM!
> 
> yeah just resolved 3.8 removals 😄

There is a test failing , but its not related to this. Would you be able to merge please?

**Pull request merged**
